### PR TITLE
Checkpoint Incentivization

### PR DESCRIPTION
**Pull request by jalextowle**
# Description

This PR implements a system that can be used to incentivize checkpoint rewards. This system has been intentionally designed to be flexible enough to adjust to new demands and to be able to be ignored entirely. 

One interesting thing that came up while testing is that rewarding the minting of past checkpoints is potentially dangerous. This is dangerous because a minter could go back and mint all of the past checkpoints to drain the subrewarder's source address.

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed. If there are
multiple reviewers, copy the checklists into sections titled `## [Reviewer Name]`.
If the PR doesn't touch Solidity, the corresponding checklist can
be removed.

## [[Reviewer Name]]

- [ ] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [ ] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [ ] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [ ] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [ ] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [ ] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### CODE REVIEW

File: contracts/src/interfaces/IHyperdriveRegistry.sol [hunk 1]
```diff
@@ -53,13 +71,36 @@ interface IHyperdriveRegistry {
         uint256 _endIndex
     ) external view returns (address[] memory);
 
+    /// @notice Gets the hyperdrive factory info for a factory.
```
Comment (jrhea) [inline contracts/src/interfaces/IHyperdriveRegistry.sol:74]:
```suggestion
    /// @notice Gets the Hyperdrive factory info for a factory.
```

File: contracts/src/interfaces/IHyperdriveRegistry.sol [hunk 2]
```diff
@@ -53,13 +71,36 @@ interface IHyperdriveRegistry {
...
+    /// @param _factory The factory.
+    /// @return The hyperdrive factory info.
```
Comment (jrhea) [inline contracts/src/interfaces/IHyperdriveRegistry.sol:76]:
```suggestion
    /// @return The Hyperdrive factory info.
```

File: contracts/src/internal/HyperdriveCheckpoint.sol [hunk 3]
```diff
@@ -315,6 +325,23 @@ abstract contract HyperdriveCheckpoint is
...
         );
 
+        // Claim the checkpoint reward on behalf of the sender.
+        //
+        // NOTE: We do this in a low-level call and ignore the status to ensure
+        // that the checkpoint will be minted regardless of whether or not the
+        // call succeeds.
+        if (_checkpointRewarder != address(0)) {
+            bool isTrader = _isTrader; // avoid stack-too-deep
+            (bool _success, ) = _checkpointRewarder.call(
```
Comment (jrhea) [inline contracts/src/internal/HyperdriveCheckpoint.sol:335]: weren't you going to wrap this in a try-catch?

File: contracts/src/rewarder/HyperdriveCheckpointRewarder.sol [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+pragma solidity 0.8.20;
+
+import { IERC20 } from "../interfaces/IERC20.sol";
+import { IHyperdriveCheckpointRewarder } from "../interfaces/IHyperdriveCheckpointRewarder.sol";
+import { IHyperdriveCheckpointSubrewarder } from "../interfaces/IHyperdriveCheckpointSubrewarder.sol";
+import { VERSION } from "../libraries/Constants.sol";
+
+/// @author DELV
+/// @notice A checkpoint rewarder that is controlled by an admin and delegates
+///         it's reward functionality to a subrewarder.
```
Comment (jrhea) [inline contracts/src/rewarder/HyperdriveCheckpointRewarder.sol:11]:
```suggestion
///         its reward functionality to a subrewarder.
```

File: contracts/src/rewarder/HyperdriveCheckpointSubrewarder.sol [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /// @notice Allows the admin to update the reward token.
+    /// @param _rewardToken The new reward token.
+    function updateRewardToken(IERC20 _rewardToken) external onlyAdmin {
+        rewardToken = _rewardToken;
+        emit RewardTokenUpdated(_rewardToken);
+    }
+
+    /// @notice Allows the admin to update the source.
```
Comment (jrhea) [inline contracts/src/rewarder/HyperdriveCheckpointSubrewarder.sol:101]: Suggestion to make it more clear:

    /// @notice Allows the admin to update the source address containing the rewards

File: test/units/rewarder/HyperdriveCheckpointRewarder.t.sol [hunk 6]
```diff
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (jrhea) [inline test/units/rewarder/HyperdriveCheckpointRewarder.t.sol:1]: Maybe name the file `MockHyperdriveCheckpointSubrewarder.t.sol` so it matched the contract name

File: contracts/src/interfaces/IHyperdriveCheckpointSubrewarder.sol [hunk 7]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.20;
+
+import { IERC20 } from "./IERC20.sol";
+import { IHyperdriveRegistry } from "../interfaces/IHyperdriveRegistry.sol";
+
+// FIXME: Add the remaining functions to this interface.
```
Comment (mcclurejt) [inline contracts/src/interfaces/IHyperdriveCheckpointSubrewarder.sol:7]: Reminder to fix

File: contracts/src/rewarder/HyperdriveCheckpointRewarder.sol [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    /// @param _subrewarder The subrewarder that processes checkpoint rewards.
+    function updateSubrewarder(
+        IHyperdriveCheckpointSubrewarder _subrewarder
+    ) external onlyAdmin {
+        subrewarder = _subrewarder;
+        emit SubrewarderUpdated(_subrewarder);
+    }
+
+    /// @notice Claims a checkpoint reward.
+    /// @param _claimant The address that is claiming the checkpoint reward.
```
Comment (mcclurejt) [inline contracts/src/rewarder/HyperdriveCheckpointRewarder.sol:65]: Why not use `@inheritdoc` here?

**Review by mcclurejt — COMMENTED** [review]
Overall LGTM, would like to see the appropriate changes made to the deployment scripting as well :)

File: contracts/src/internal/HyperdriveCheckpoint.sol [hunk 9]
```diff
@@ -315,6 +325,23 @@ abstract contract HyperdriveCheckpoint is
...
+                abi.encodeCall(
+                    IHyperdriveCheckpointRewarder.claimCheckpointReward,
+                    (msg.sender, checkpointTime, isTrader)
+                )
+            );
```
Comment (Saw-mon-and-Natalie) [inline contracts/src/internal/HyperdriveCheckpoint.sol:340]: `returndata` bombing is possible here to cause OOG in `hyperdrive`:
- https://github.com/spearbit-audits/delv-week-review/issues/2

File: contracts/src/internal/HyperdriveCheckpoint.sol (hunk 3 above)
Comment (jalextowle) [reply contracts/src/internal/HyperdriveCheckpoint.sol:335]: `try-catch` doesn't catch all errors. This issue goes into some details: https://github.com/ethereum/solidity/issues/12725. The issue mentioned in the original post doesn't pertain to this, but there are other issues that can also result in reverts. Doing a low-level call is (unfortunately) safer.

File: test/units/rewarder/HyperdriveCheckpointRewarder.t.sol (hunk 6 above)
Comment (jalextowle) [reply test/units/rewarder/HyperdriveCheckpointRewarder.t.sol:1]: The file is named after the test contract (the second contract defined here). This is also the pattern used in the reentrancy tests FWIW

File: contracts/src/interfaces/IHyperdriveRegistry.sol [hunk 10]
```diff
@@ -53,13 +71,36 @@ interface IHyperdriveRegistry {
...
+    /// @param _factory The factory.
+    /// @return The Hyperdrive factory info with associated metadata.
+    function getFactoryInfoWithMetadata(
+        address _factory
+    ) external view returns (FactoryInfoWithMetadata memory);
+
+    /// @notice Gets the Hyperdrive factory info with associated metadata for a
+    ///         list of factories.
+    /// @param __factories The list of factories.
+    /// @return The hyperdrive factory info with associated metadata.
```
Comment (jrhea) [inline contracts/src/interfaces/IHyperdriveRegistry.sol:99]: This should indicate that a list is returned or at least use the word infos

File: contracts/src/interfaces/IHyperdriveRegistry.sol [hunk 11]
```diff
@@ -79,10 +120,32 @@ interface IHyperdriveRegistry {
...
+    /// @param _instance The instance.
+    /// @return The instance info with associated metadata.
+    function getInstanceInfoWithMetadata(
+        address _instance
+    ) external view returns (InstanceInfoWithMetadata memory);
+
+    /// @notice Gets the instance info with associated metadata for a list of
+    ///         instances.
+    /// @param __instances The list of instances.
+    /// @return The instance info with associated metadata.
```
Comment (jrhea) [inline contracts/src/interfaces/IHyperdriveRegistry.sol:147]: should indicate that a list is returned

File: contracts/src/internal/HyperdriveCheckpoint.sol [hunk 12]
```diff
@@ -315,6 +327,27 @@ abstract contract HyperdriveCheckpoint is
...
 
+        // Claim the checkpoint reward on behalf of the sender.
+        //
+        // NOTE: We do this in a low-level call and ignore the status to ensure
+        // that the checkpoint will be minted regardless of whether or not the
+        // call succeeds. Furthermore, we use the `ExcessivelySafeCall` library
+        // to prevent returndata bombing.
+        if (_checkpointRewarder != address(0)) {
+            bool isTrader = _isTrader; // avoid stack-too-deep
+            (bool _success, ) = _checkpointRewarder.excessivelySafeCall(
```
Comment (jrhea) [inline contracts/src/internal/HyperdriveCheckpoint.sol:338]: can we just get rid of this assignment?

**Review by jrhea — APPROVED** [review]
lgtm. just a couple of nits. one thing to note, i had to eyeball the coverage bc coveralls diffs were timing out on me.

**Pull request merged**